### PR TITLE
Generate types and type specs for all generated functions in aws-elixir and aws-erlang

### PR DESCRIPTION
**Pull request by onno-vos-dev**
### Description
Generate types and type specs for all generated functions in aws-elixir and aws-erlang

- Generate common errors in order to unify types and reduce noise
- Ensure that for Elixir, typedoc examples are considered code blocks
- Erlang Type docs are not hoverable for now due to issues with 'rebar3 ex_doc'
  - Will follow up on this task
- Increase Task.await/2 timeout

### Follow up tasks
Tasks to follow up on:
- [ ] `aws_client()` in `aws-erlang` is marked as `opaque` but this is not really true due to `build_host/2`. We should either fix this to be truly `opaque` or simply turn it into a regular `type`. 
  - [ ] Due to this Dialyzer is not really happy with `aws-erlang` and while getting it down to 0 is likely gonna be a challenge, solving the above bring it down to 27 which is a lot more reasonable. 
- [x] Add `nowarn_unused_type` to `rebar.config` in `aws-erlang` since some types are only used from comments (examples) and hence are "unused". Let's not flood the terminal with noise. 
  - [x]  PR: https://github.com/aws-beam/aws-erlang/pull/151

### Generated code
Generated aws-elixir code as an example: [comparison](https://github.com/aws-beam/aws-elixir/compare/master-snapshot-20240313...add-typespecs-to-all-generated-functions?expand=1)
Generated aws-erlang code as an example: [comparison](https://github.com/aws-beam/aws-erlang/compare/master-snapshot-20240313...add-typespecs-to-all-generated-functions?expand=1)

### Additional info
Related to #108 which I declined in favor of this one due to wrong branch naming. 

#### ElixirLS overview
Screenshot 2024-03-13 at 20 58 20

#### Erlang types overview
Screenshot 2024-03-16 at 11 45 27
Screenshot 2024-03-16 at 11 45 48

### CODE REVIEW

File: priv/post.ex.eex [hunk 1]
```diff
@@ -11,6 +11,46 @@ defmodule <%= context.module_name %> do
...
+    <% end %>
+"""
+@type <%= if map_size(type_fields) == 0 do "#{type_name}() :: %{}" else "#{type_name}() :: %{String.t => any()}" end %>
+<% end %>
+
+<%= Enum.map(context.actions,
+             fn action ->
+                errors = action.errors
+                if not is_nil(errors) do
+                    errors_snakecased = errors |> Enum.map(fn error -> AWS.CodeGen.Name.to_snake_case(String.replace(error["target"], ~r/com\.amazonaws\.[^#]+#/, "")) end)
```
Comment (philss) [inline priv/post.ex.eex:35]: It's just style, so don't bother, but we usually prefer to not use a pipe when it's just one function in the context.

```suggestion
                    errors_snakecased = Enum.map(errors, fn error -> AWS.CodeGen.Name.to_snake_case(String.replace(error["target"], ~r/com\.amazonaws\.[^#]+#/, "")) end)
```

File: priv/rest.erl.eex [hunk 2]
```diff
@@ -8,19 +8,60 @@
 
 -include_lib("hackney/include/hackney_lib.hrl").
 
+<%= for {type_name, type_fields} <- AWS.CodeGen.Types.types(context) do %>
+<%= if Map.keys(type_fields) == [] do %>%% Example:
```
Comment (philss) [inline priv/rest.erl.eex:12]: You can check if the Enum is empty here:

```suggestion
<%= if Enum.empty?(type_fields) do %>%% Example:
```

File: priv/rest.ex.eex [hunk 3]
```diff
@@ -11,6 +11,46 @@ defmodule <%= context.module_name %> do
...
+    <%= "#{type_name}() :: %{}" %>
+<% else %>
+    <%= "#{type_name}() :: %{" %>
+    <%= Enum.map_join(type_fields, ",\n    ", fn {field_name, field_type} ->
+      ~s{  #{field_name}#{field_type}}
+    end) %>
+    }
+<% end %>
+"""
+@type <%= if map_size(type_fields) == 0 do "#{type_name}() :: %{}" else "#{type_name}() :: %{String.t => any()}" end %>
```
Comment (philss) [inline priv/rest.ex.eex:28]: Same here: you can use `Enum.empty?/1`.

```suggestion
@type <%= if Enum.empty?(type_fields) do "#{type_name}() :: %{}" else "#{type_name}() :: %{String.t => any()}" end %>
```

File: priv/rest.ex.eex [hunk 4]
```diff
@@ -11,6 +11,46 @@ defmodule <%= context.module_name %> do
   alias AWS.Client
   alias AWS.Request
 
+ <%= for {type_name, type_fields} <- AWS.CodeGen.Types.types(context) do %>
+@typedoc """
+
+## Example:
+<%= if map_size(type_fields) == 0 do %>
```
Comment (philss) [inline priv/rest.ex.eex:18]:
```suggestion
<%= if Enum.empty?(type_fields) do %>
```

File: test/aws_codegen_test.exs [hunk 5]
```diff
@@ -69,105 +69,154 @@ defmodule AWS.CodeGenTest do
...
 
-                 @doc \"\"\"
-                 Ingests your application events into CloudTrail Lake.
+                 ## Example:
+
+                     audit_event() :: %{
+                       \"eventData\" => [String.t()],
+                       \"eventDataChecksum\" => [String.t()],
+                       \"id\" => String.t()
+                     }
```
Comment (philss) [inline test/aws_codegen_test.exs:80]: Beautiful! 😍

**Review by philss — APPROVED** [review]
Awesome work, @onno-vos-dev!! 🎉 ❤️ 

I liked very much the way it is describing the type! I hope in the future we can include the entire type description in the type specs.

I added some minor points only, but consider it approved! 😺 
Please don't forget to run the formatter :)

File: test/aws_codegen_test.exs (hunk 5 above)
Comment (onno-vos-dev) [reply test/aws_codegen_test.exs:80]: ❤️😍😍😍

File: lib/aws_codegen/shapes.ex [hunk 6]
```diff
@@ -9,6 +19,23 @@ defmodule AWS.CodeGen.Shapes do
     get_in(operation_spec, ["output", "target"])
   end
 
+  def collect_shapes(_language, api_spec) do
+    api_spec["shapes"]
+    |> Enum.sort(fn {name_a, _}, {name_b, _} -> name_a < name_b end)
```
[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply lib/aws_codegen/shapes.ex:24]: Solved

File: lib/aws_codegen/types.ex [hunk 7]
```diff
@@ -0,0 +1,290 @@
+defmodule AWS.CodeGen.Types do
+
+  alias AWS.CodeGen.Shapes.Shape
+
+  def types(context) do
+    Enum.reduce(context.shapes, Map.new(), fn {_name, shape}, acc ->
```
[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply lib/aws_codegen/types.ex:6]: Fixed in both places in this module 👍

**Pull request merged**